### ISSUE TITLE
S3: cannot copy file bigger than 8MB  to S3

### ISSUE THREAD

**Issue by martin-voj-se** [localstack/localstack#336]
I've installed localstack on my laptop: Ubuntu 16.04, 16GB RAM, Python 2.7.
pip install localstack

localstack (0.8.0)
localstack-ext (0.7.0.7)
boto (2.46.1)
boto3 (1.4.4)
botocore (1.5.49)
(...)

I can start it, create bucket, copy small file, ls bucket:

```
$ localstack start
Starting local dev environment. CTRL-C to quit.
Starting mock API Gateway (http port 4567)...
Starting mock CloudWatch (http port 4582)...
Starting mock SES (http port 4579)...
Starting mock Kinesis (http port 4568)...
Starting mock Redshift (http port 4577)...
Starting mock S3 (http port 4572)...
Starting mock CloudFormation (http port 4581)...
Starting mock DynamoDB (http port 4569)...
Starting mock SQS (http port 4576)...
Starting local Elasticsearch (http port 4571)...
Starting mock SNS (http port 4575)...
Starting mock DynamoDB Streams service (http port 4570)...
Starting mock Firehose service (http port 4573)...
Starting mock Route53 (http port 4580)...
Starting mock ES service (http port 4578)...
Starting mock Lambda service (http port 4574)...
Ready.
```

```
aws --endpoint-url=http://localhost:4572 s3 mb s3://new_bucket
aws --endpoint-url=http://localhost:4572 s3 ls
aws --endpoint-url=http://localhost:4572 s3 cp /home/martin/data/wrk-aws/pic.png s3://new_bucket

aws --endpoint-url=http://localhost:4572 s3 ls s3://new_bucket/
2017-09-20 16:12:04     117563 pic.png
```

I can copy 8MB file but not 9MB file:

```
-rw-rw-r-- 1 martin martin   8192000 Sep 20 15:44 data.8MB.raw
-rw-rw-r-- 1 martin martin   9216000 Sep 20 15:52 data.9MB.raw

aws --endpoint-url=http://localhost:4572 s3 cp /home/martin/data/wrk-aws/data.9MB.raw s3://new_bucket

[error msgs in terminal]

2017-09-20T16:12:51:ERROR:localstack.services.generic_proxy: Error forwarding request: cannot concatenate 'str' and 'NoneType' objects Traceback (most recent call last):
  File "/home/martin/.local/lib/python2.7/site-packages/localstack/services/generic_proxy.py", line 171, in forward
    path=path, data=data, headers=forward_headers)
  File "/home/martin/.local/lib/python2.7/site-packages/localstack/services/s3/s3_listener.py", line 345, in forward_request
    expanded_data = expand_multipart_filename(original_data, headers)
  File "/home/martin/.local/lib/python2.7/site-packages/localstack/services/s3/s3_listener.py", line 260, in expand_multipart_filename
    _, params = cgi.parse_header(headers.get('Content-Type'))
  File "/usr/lib/python2.7/cgi.py", line 309, in parse_header
    parts = _parseparam(';' + line)
TypeError: cannot concatenate 'str' and 'NoneType' objects

2017-09-20T16:13:51:ERROR:localstack.services.generic_proxy: Error forwarding request: cannot concatenate 'str' and 'NoneType' objects Traceback (most recent call last):
  File "/home/martin/.local/lib/python2.7/site-packages/localstack/services/generic_proxy.py", line 171, in forward
    path=path, data=data, headers=forward_headers)
  File "/home/martin/.local/lib/python2.7/site-packages/localstack/services/s3/s3_listener.py", line 345, in forward_request
    expanded_data = expand_multipart_filename(original_data, headers)
  File "/home/martin/.local/lib/python2.7/site-packages/localstack/services/s3/s3_listener.py", line 260, in expand_multipart_filename
    _, params = cgi.parse_header(headers.get('Content-Type'))
  File "/usr/lib/python2.7/cgi.py", line 309, in parse_header
    parts = _parseparam(';' + line)
TypeError: cannot concatenate 'str' and 'NoneType' objects

2017-09-20T16:14:53:ERROR:localstack.services.generic_proxy: Error forwarding request: cannot concatenate 'str' and 'NoneType' objects Traceback (most recent call last):
  File "/home/martin/.local/lib/python2.7/site-packages/localstack/services/generic_proxy.py", line 171, in forward
    path=path, data=data, headers=forward_headers)
  File "/home/martin/.local/lib/python2.7/site-packages/localstack/services/s3/s3_listener.py", line 345, in forward_request
    expanded_data = expand_multipart_filename(original_data, headers)
  File "/home/martin/.local/lib/python2.7/site-packages/localstack/services/s3/s3_listener.py", line 260, in expand_multipart_filename
    _, params = cgi.parse_header(headers.get('Content-Type'))
  File "/usr/lib/python2.7/cgi.py", line 309, in parse_header
    parts = _parseparam(';' + line)
TypeError: cannot concatenate 'str' and 'NoneType' objects

2017-09-20T16:15:57:ERROR:localstack.services.generic_proxy: Error forwarding request: cannot concatenate 'str' and 'NoneType' objects Traceback (most recent call last):
  File "/home/martin/.local/lib/python2.7/site-packages/localstack/services/generic_proxy.py", line 171, in forward
    path=path, data=data, headers=forward_headers)
  File "/home/martin/.local/lib/python2.7/site-packages/localstack/services/s3/s3_listener.py", line 345, in forward_request
    expanded_data = expand_multipart_filename(original_data, headers)
  File "/home/martin/.local/lib/python2.7/site-packages/localstack/services/s3/s3_listener.py", line 260, in expand_multipart_filename
    _, params = cgi.parse_header(headers.get('Content-Type'))
  File "/usr/lib/python2.7/cgi.py", line 309, in parse_header
    parts = _parseparam(';' + line)
TypeError: cannot concatenate 'str' and 'NoneType' objects

2017-09-20T16:17:04:ERROR:localstack.services.generic_proxy: Error forwarding request: cannot concatenate 'str' and 'NoneType' objects Traceback (most recent call last):
  File "/home/martin/.local/lib/python2.7/site-packages/localstack/services/generic_proxy.py", line 171, in forward
    path=path, data=data, headers=forward_headers)
  File "/home/martin/.local/lib/python2.7/site-packages/localstack/services/s3/s3_listener.py", line 345, in forward_request
    expanded_data = expand_multipart_filename(original_data, headers)
  File "/home/martin/.local/lib/python2.7/site-packages/localstack/services/s3/s3_listener.py", line 260, in expand_multipart_filename
    _, params = cgi.parse_header(headers.get('Content-Type'))
  File "/usr/lib/python2.7/cgi.py", line 309, in parse_header
    parts = _parseparam(';' + line)
TypeError: cannot concatenate 'str' and 'NoneType' objects

```

Can I set bigger limit or is it a bug?
Thanks.

**Comment by stanjer** [comment]
I've had this problem as well. As far as I see from aws-cli code (https://github.com/aws/aws-cli/blob/develop/awscli/customizations/s3/transferconfig.py#L21), it's sending data with POST instead of PUT in case a file is bigger than ~8Mb. The current implementation of such requests in localstack looks like broken. I was able to partly fix it locally, maybe I'll be able to provide a patch later next week, as the tests are needed, and I don't have time for it now.

**Comment by bogdanbrindusan** [comment]
@stanjer I've recently hit into same issue, any updates on getting this fixed?

**Comment by i-aggarwal** [comment]
I have hit the same issue as well. Any news on a fix or any knowledge of a workaround would be very helpful.

Thanks.

**Comment by whummer** [comment]
This should be fixed in #1200 - can you please give it a try with the latest version on `master`, or the `localstack/localstack:latest` Docker image?

Please re-open this issue if the problem persists @i-aggarwal . Thanks
